src/main/java/org/launchcode/water_garden_tour/models/garden/GardenData.java

```diff
@@ -0,0 +1,102 @@
+package org.launchcode.water_garden_tour.models.garden;
+
+import org.springframework.security.core.parameters.P;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class GardenData {
+
+    public static List<Garden> gardenSearchFeaturesFiltered(String searchValueRaw, Integer[] featureIds, Iterable<Feature> allFeatures, List<Garden> allGardens) {
+        String searchValue = searchValueRaw.toLowerCase();
+        List<Garden> results = new ArrayList<>();
+        Feature curFeature;
+
+        for (Garden garden : allGardens) {
+            boolean hasAllFeatures = true;
+
+            for (Integer featureId : featureIds) {
+                curFeature = getFeatureFromId(featureId, allFeatures);
+
+                if (garden.getFeatures().contains(curFeature)) {
+                } else {
+                    hasAllFeatures = false;
+                }
+            }
+            if (hasAllFeatures) {
+                results.add(garden);
+            }
+        }
+
+        if (searchValue.equals("all") || searchValue.equals("")){
+
+        } else {
+            List<Garden> iteratingGardenList = new ArrayList<>();
+            iteratingGardenList.addAll(results);
+
+            for (Garden garden : iteratingGardenList) {
+                if (!isTextInGardenFields(garden, searchValue)) {
+                    results.remove(garden);
+                }
+            }
+
+        }
+        return results;
+    }
+
+    public static List<Garden> gardenSearchAllFeatures(String searchValueRaw, Iterable<Feature> allFeatures, List<Garden> allGardens) {
+        String searchValue = searchValueRaw.toLowerCase();
+        List<Garden> results = new ArrayList<>();
+
+        if (searchValue.equals("all") || searchValue.equals("")){
+            results = allGardens;
+        } else {
+
+        }
+            for (Garden garden : allGardens) {
+
+                for(Feature feature : garden.getFeatures()) {
+
+                    if (feature.getName().toLowerCase().contains(searchValue)) {
+
+                        if (!results.contains(garden)) {
+
+                            results.add(garden);
+                        }
+                    }
+                }
+
+                if (isTextInGardenFields(garden, searchValue)) {
+                    if (!results.contains(garden)) {
+                        results.add(garden);
+                    }
+                }
+        }
+
+        return results;
+    }
+
+    public static Feature getFeatureFromId(Integer featureId, Iterable<Feature> allFeatures) {
+
+        for (Feature feature : allFeatures) {
+            if (featureId.equals(feature.getId())) {
+                return feature;
+            }
+        }
+        return null;
```

Comment:
no reason to do anything here but be careful when returning null values as when you call this in the future you might get null pointers and think "Hey my objects are empty or not holding anything"

Reply:
That makes sense, duly noted